File: solutions/ruby/triangle/1/triangle.rb
class Triangle
  def initialize(tri)
    @a, @b, @c = *tri
  end

  def equilateral?
    !a.zero? &&
      (a == b && b == c)
  end

  def isosceles?
    !a.zero? &&
      inequality? && (a == b || b == c || a == c)
  end

  def scalene?
    inequality? && (a != b && b != c)
  end

private
  attr_reader :a, :b, :c

  def inequality?
    a + b >= c &&
      a + c >= b &&
      b + c >= a
  end
end

File: solutions/clojure/triangle/1/src/triangle.clj
(ns triangle)

(defn- pre?
  [a b c]
  (and (not (or (zero? a) (zero? b) (zero? c)))
       (>= (+ a c) b)
       (>= (+ a b) c)
       (>= (+ c b) a)))
(defn equilateral? [a b c] (and (pre? a b c) (== a b c)))
(defn isosceles? [a b c] (and (pre? a b c) (or (== a b) (== b c) (== c a))))
(defn scalene? [a b c] (and (pre? a b c) (not= a b) (not= b c) (not= a c)))
